code/game/objects/machinery/miner.dm

```diff
@@ -43,21 +50,90 @@
 /obj/machinery/miner/update_icon()
 	switch(miner_status)
 		if(MINER_RUNNING)
-			icon_state = "mining_drill_active"
+			if((mineral_produced == /obj/item/compactorebox/platinum)&&(miner_upgrade_type == MINER_COMPACTOR))
+				icon_state = "mining_drill_active_platinum_[miner_upgrade_type]"
+			else
+				icon_state = "mining_drill_active_[miner_upgrade_type]"
 		if(MINER_SMALL_DAMAGE)
-			icon_state = "mining_drill_braced"
+			icon_state = "mining_drill_braced_[miner_upgrade_type]"
 		if(MINER_MEDIUM_DAMAGE)
-			icon_state = "mining_drill"
+			icon_state = "mining_drill_[miner_upgrade_type]"
 		if(MINER_DESTROYED)
-			icon_state = "mining_drill_error"
+			icon_state = "mining_drill_error_[miner_upgrade_type]"
+
+/obj/machinery/miner/proc/attempt_upgrade(obj/item/minerupgrade/upgrade, mob/user, params)
+	if(miner_upgrade_type)
+		return FALSE
+	if(user.skills.getRating("construction")<SKILL_CONSTRUCTION_ADVANCED)
+		to_chat(user, "<span class='info'>You can't figure out how to install the complex module.</span>")
+		return FALSE
+	user.visible_message("<span class='notice'>[user] begins attaching a module to [src]'s sockets.</span>")
+	to_chat(user, "<span class='info'>You begin installing the [upgrade] on the miner.</span>")
+	if(!do_after(user, 15 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return FALSE
+	if(!upgrade.use())
+		return FALSE
+	var/choice = upgrade.uptype
```

Comment:
No need for this var just use upgrade.uptype

Reply:
done.